### ISSUE TITLE
Should objects expose the descriptor they were created with?

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#1498]
Today, there's no way to do "reflection" on an object to see what descriptor it was created with (or what the results of defaulting were, or see what fields were obeyed), with a few exceptions: `GPUDevice.features`, `GPUDevice.limits`, and `GPUPipelineBase.getBindGroupLayout()`.

Should we reflect descriptors for other objects?
- Easy cases: buffer, texture, etc.
- Hard cases: bind group layout, bind group

See also: #1497, #1309

**Comment by AshleyScirra** [comment]
Just mentioning here that as described in #1543, it would be convenient to have these for library methods. For example a library method to convert the format of a texture would require inspecting the source texture's usage, format, dimensions and size.

**Comment by almarklein** [comment]
As I undestand it, this proposal is to makes developers lives easier by exposing things like a buffer's and texture's size and usage, and the descriptors provides that info (and more). However, being able to query e.g. `GPUBufferDescriptor.mappedAtCreation` after it being instantiated seems a bit strange. Plus it costs memory to store these data.

In many cases the info that would be useful to expose is an intrinsic feature of the object and already present as an internal slot. It would be easy for implementations to expose that info at no extra cost in terms of memory. This would probably not cover *all* info that is potentially useful, but it could be something to be consistent about: only exposing data that is already available internally. Just a thought.

**Comment by kvark** [comment]
@almarklein there are at least 2 issues with this:
  1. what is available internally is an implementation detail. It may need to change. At the same time, Apple is going to make an implementation and realize that these APIs would constrain it, for example.
  2. the info is available internally in the GPU process, but this issue is about exposing it on the content side. So if we follow this, we'd have to make those APIs asynchronous, which would make them much less useful.

**Comment by almarklein** [comment]
Thanks for the clarification. Out of interest, if these internal values are implementation details, why are they specified in the WebGPU spec (as internal slots)?

**Comment by Kangz** [comment]
Because the specification needs to use the data from the descriptor to describe what validation is performed on GPU operations. However most of the validation happens on the "device timeline" which means in the GPU process.

**Comment by kainino0x** [comment]
> As I undestand it, this proposal is to makes developers lives easier by exposing things like a buffer's and texture's size and usage, and the descriptors provides that info (and more). However, being able to query e.g. `GPUBufferDescriptor.mappedAtCreation` after it being instantiated seems a bit strange. Plus it costs memory to store these data.

This makes sense to me, we should probably limit the reflected info to just specific things like size, format, usage, etc.

**Comment by Kangz** [comment]
Probably want to discuss this for v1.

**Comment by brendan-duncan** [comment]
One more useful bit of reflection is the mapped state of a buffer: unmapped, pending map, mapped.

**Comment by Kangz** [comment]
Exposing additional creation parameters is a non-breaking change, so I think we should be somewhat conservative in what we expose in WebGPU v1 to limit implementation work and overhead. However two pieces of information seem most useful to expose:

 - The state of resources that have state, especially `GPUBuffer` (the other only have valid vs. destroyed). We'll need to be careful about what to do for the "invalid" state that cannot be 100% known client side.
 - The creation parameters of things that are shared between parts of a program to help make WebGPU more composable. I think this means most of the creation parameters of `GPUBuffer` `GPUTexture` and `GPUQuerySet` because they are containers and it is useful for libraries etc to know the size / format of these containers. `GPUSampler` is less critical because it contains a fixed amount of parameters given at creation.

So the proposal is the following:

```webidl
partial interface GPUQuerySet {
    readonly GPUQueryType type;
    readonly GPUSize32 count;
}
partial interface GPUTexture {
    readonly GPUExtent3D size;
    readonly GPUIntegerCoordinate mipLevelCount;
    readonly GPUSize32 sampleCount;
    readonly GPUTextureDimension dimension;
    readonly GPUTextureFormat format;
    readonly GPUTextureUsageFlags usage;
}

enum GPUBufferState {
    "unmapped",
    "mapped",
    "map pending",
    "destroyed",
}
partial interface GPUBuffer {
    readonly GPUSize64 size;
    readonly GPUBufferUsageFlags usage;

    readonly GPUBufferState state;
}
```

**Comment by kvark** [comment]
Exposing this for query sets, textures, and buffers, while omitting for bind groups, pipeline layouts, texture views, and potentially other things... means the line is drawn arbitrarily. Either have a strong explanation on why this subset is chosen, or not expose this at all. Like @litherum mentioned in https://github.com/gpuweb/gpuweb/issues/1497#issuecomment-996974161, it's trivial to wrap objects in JS and keep the descriptors around. Could be done by user, or by a middle-layer.

**Comment by Kangz** [comment]
> the line is drawn arbitrarily

Yes and no, I tried to give a rational of why the line is there: it is the sizing of container resources because that's what's most likely to be useful for the composability of applications. (at least from my intuition, which I agree is not a scientific measure)

Would you draw the line somewhere else or just not expose anything?

**Comment by shannon** [comment]
I just wanted to provide one example that I came across that required passing around a descriptor that seemed a bit unwieldy. Because WebGPU doesn't provide a method for generating mipmaps you have to implement this in application space. This requires knowing the texture format and mipLevelCount after the texture exists. 

See the example here
https://github.com/toji/webgpu-best-practices/blob/main/img-textures.md#generating-mipmaps

The same can be said implementing a readTexture method. So it seems to me that in the case of missing common functionality in WebGPU, not having this information exposed makes it all the more complicated to add in application space.

**Comment by kainino0x** [comment]
Similar to by thoughts on https://github.com/gpuweb/gpuweb/issues/1497#issuecomment-1006930846 the longer this issue is open the more I want to just get it over with. It's clear there are a bunch of cases where this is convenient even if not strictly necessary, and I'm not concerned about implementation or performance costs. The hard part is really just where to draw the line, but I don't think it's a big deal if we draw the line arbitrarily.

**Comment by AshleyScirra** [comment]
> it's trivial to wrap objects in JS and keep the descriptors around

I don't think this is a good solution for library methods, e.g. for generating mipmaps, converting textures etc. If every app rolls its own JS wrappers, then there is no standard way for library methods to find parameters like texture format and mip level count. These would probably just have to all be passed as extra parameters, shunting the problem of how to track them back on to the caller; and suddenly discovering you have to wrap all your WebGPU objects in JS wrappers can mean suddenly having to do a load of refactoring.

I think exposing only creation parameters for immutable objects should be straightforward and uncontroversial, and largely solve the problem for libraries, which will start popping up as soon as v1 is out.

**Comment by jdashg** [comment]
Having written a bunch of wrapping layers myself, I wish I didn't have to.
Exposing this data is cheap-to-free for us, and practically speaking doesn't even need to be validated.

**Comment by Kangz** [comment]
I'd like to suggest that we only add the shape of container objects for v1 (buffer, texture, queries) and leave the rest to post-v1 polish.

**Comment by kainino0x** [comment]
From #2919:

> I'm thinking we won't expose the whole descriptor. If some application like Spector.js really wants the whole descriptor, they should probably just do the prototype-shim thing and save the descriptor off themselves. Spector.js is going to want to save off a bunch of information for better debuggability that we're never going to want to save, for example entire bind group descriptors, pipeline descriptors, and probably even the contents of command buffers. So we're never going to be able to satisfy all of the needs of Spector.js ourselves.
> 
> What we _should_ focus on satisfying is the convenience of helper functions/libraries/middleware, which is exactly why we're adding these bits of reflection. And those don't care if we flatten the thing or not.

**Comment by greggman** [comment]
Above there are comments on exposing the state of mapping of buffers. Is that still on the table? Does that have any implications for multi-threading?

**Comment by Kangz** [comment]
I think we could have it still on the table for v1, up for discussion with the group. My only worry is that it will make the wording and transitions of the spec developer-visible so we won't be able to change them afterwards (it would be a backwards incompatible change)

**Comment by kainino0x** [comment]
I'm not too worried about that since I think all the states are essentially observable/trackable already anyway (see #3010) but I'm also not sure we need more than that for 1.0 (I filed that because it would be breaking change for 1.0).

As for multithreading, there are some details in how the state would be exposed exactly, but again, even without us exposing the state directly, it's mostly observable (mapAsync will fail if already mapped/pending on another thread) and completely trackable (can be tracked in a SharedArrayBuffer). This is why the spec requires site isolation for multithreaded usage: https://gpuweb.github.io/gpuweb/#security-timing

**Comment by Kangz** [comment]
For reference reflection for container objects was added to the spec in #2919. (I forgot to reference this issue in the PR commit message).
